### PR TITLE
fix #4054

### LINKED ISSUE #4054: Spacer hangs when unused (unconstrained) var is added
Given these two queries:
1. https://gist.github.com/leonardoalt/0cee9cb2e86082b9c7f2fa1c7c88a0be
2. https://gist.github.com/leonardoalt/aa46fc20d937b8d85258a08e10f1cd32

The second one is basically the first one plus an `Array` variable that is just passed around the rules but unconstrained.
The first one is solved in 0.4s, whereas the second one hangs.

Any idea why that happens?

### PR DESCRIPTION
**Pull request by hgvk94**

### CODE REVIEW

File: src/muz/spacer/spacer_util.cpp [hunk 1]
```diff
@@ -980,7 +980,13 @@ namespace {
         for_each_expr(cd, fml);
     }
 
-}
-
+    // set the value of a boolean function to true in model
+    void set_true_in_mdl(model *model, func_decl *f) {
```
Comment (NikolajBjorner) [inline src/muz/spacer/spacer_util.cpp:984]: pass as model& model instead of model* model

File: src/muz/spacer/spacer_context.cpp [hunk 2]
```diff
@@ -3321,6 +3347,96 @@ void context::predecessor_eh()
...
+  }
+  return true;
+}
+
+// Handle cases where solver returns unknown but returns a good enough model
+// model is good enough if it satisfies
+// 1. all the reachable states whose tag is set in the model
+// 2. Tr && pob
+lbool context::handle_unknown(pob &n, const datalog::rule *r, model_ref model) {
+  if (r == nullptr) {
```
Comment (NikolajBjorner) [inline src/muz/spacer/spacer_context.cpp:3426]: 4 spaces
https://www.bing.com/videos/search?q=tabs+vs+spaces&&view=detail&mid=D8125B650297896DCD76D8125B650297896DCD76&&FORM=VRDGAR

Comment (hgvk94) [reply src/muz/spacer/spacer_context.cpp:3426]: But .... I used clang format.

Comment (NikolajBjorner) [reply src/muz/spacer/spacer_context.cpp]: it is configurable

**Comment by NikolajBjorner** [comment]
@agurfinkel - are you handling the merge status of this pull request?

**Comment by agurfinkel** [comment]
@NikolajBjorner yes. waiting for spaces to normalize and builds to pass

**Comment by agurfinkel** [comment]
@hgvk94 the build issue seems to be related to something outside of this PR. I can reproduce it (in release build only) on your PR, but not any longer on current `master`. 

Please rebase against master and re-submit to trigger CI one more time.

**Comment by agurfinkel** [comment]
@NikolajBjorner is there some known instability in the build right now? This PR keeps failing regressions, but, as far as I can tell, no line of code from the PR is actually executed in the failing test.

Previously, failure was on a mac in seq theory initialization. I could reproduce it locally. Now a string benchmark in timing out on some linux build. I can dig deeper, but I wanted to make sure that there isn't a known anomaly.

**Comment by NikolajBjorner** [comment]
string-ops.smt2 started timing out yesterday. Seems a butterfly effect of slow string reasoning

**Comment by agurfinkel** [comment]
then this PR is ready to go. We can wait until the CI is more stable and re-submit. Let me know what you want to do.

**Pull request merged**